src/PieChart.php

```diff
@@ -4,64 +4,74 @@
-    public $type = 'pie';
+    public string $type = self::TYPE_PIE;
 
-    public function setModel(Model $model, array $columns = [], array $stacks = []): void
+    public function prepareDatasets(): void
     {
-        if ($columns === []) {
-            throw new Exception('Second argument must be specified to Chart::setModel()');
-        }
-
-        $this->datasets = [];
+        $datasets = [];
         $colors = [];
 
         // initialize data-sets
-        foreach ($columns as $key => $column) {
-            $colors[$column] = $this->niceColors;
-
+        foreach ($this->columns as $key => $column) {
             if ($key === 0) {
                 $titleColumn = $column;
 
-                continue; // skipping labels
+                continue; // skipping label column
             }
 
-            $this->datasets[$column] = [
+            $colors[$column] = new Color();
+
+            $datasets[$column] = [
                 'data' => [],
+                'label' => $this->model->getField($column)->getCaption(),
                 'backgroundColor' => [],
+                'borderColor' => [],
+                'borderWidth' => 1,
+                'hoverOffset' => 8,
+                'borderAlign' => 'inner',
             ];
         }
 
         // prepopulate data-sets
-        foreach ($model as $entity) {
+        foreach ($this->model as $entity) {
             $this->labels[] = $entity->get($titleColumn); // @phpstan-ignore-line
-            foreach ($this->datasets as $key => &$dataset) {
-                $dataset['data'][] = $entity->get($key);
-                $color = array_shift($colors[$key]);
-                $dataset['backgroundColor'][] = $color[0];
-                $dataset['borderColor'][] = $color[1];
+            foreach ($datasets as $column => $dataset) {
+                $datasets[$column]['data'][] = $entity->get($column);
+                $color = $colors[$column]->getNextColorPair();
+                $datasets[$column]['backgroundColor'][] = $color[0];
+                $datasets[$column]['borderColor'][] = $color[1];
             }
         }
+
+        $this->setDatasets($datasets);
     }
 
-    public function withCurrency(string $char = '€', string $axis = 'y')
+    public function setCurrencyLabel(string $char = '€', string $axis = 'y', int $digits = 2)
     {
-        $options = [];
-        $options['tooltips'] = [
-            'callbacks' => [
-                'label' => new JsExpression('{}', [
-                    'function(item, data, bb) {
-                        var val = data.datasets[item.datasetIndex].data[item.index];
-
-                        return "' . $char . '" +  val.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ",");
-                    }',
-                ]),
+        $options = [
+            'plugins' => [
+                'tooltip' => [
+                    'enabled' => true,
+                    'mode' => 'point',
+                    'callbacks' => [
```

Comment:
in atk4/* we prefer as much as possible things to be done in php, here format the values using atk4/ui UI persistence - can we do that?

Reply:
Maybe we can with some workaround (like first storing all labels as JS array filled from PHP side) and then in JS callback lookup that value, but I don't think it's worth it. This way it's easy and straightforward at least for now.
There definitely is space for improvement - maybe someone who use charts in projects will contribute something nicer with a time, but for now ... i guess it's fine like this.